mhw-cs-plugin-loader/Preloader.cpp

```diff
@@ -53,26 +55,30 @@ void OpenConsole() {
 // This runs before all of the CRT initalization, static initalizers, and WinMain.
 __declspec(noinline) int64_t hookedSCRTCommonMain()
 {
-    dlog::debug("Start initializing CLR/NativePluginFramework (before static-initalizers)");
-
+    dlog::info("Initializing CLR / NativePluginFramework");
     s_coreclr = new CoreClr();
     s_framework = new NativePluginFramework(s_coreclr);
     dlog::info("Initialized");
 
-    // Trigger plugin preload event
-    s_framework->PreloadPlugins();
+    s_framework->TriggerOnPreMain();
 
     return g_SCRTCommonMain_hook.call<int64_t>();
 }
 
 __declspec(noinline) int __stdcall hookedWinMain(HINSTANCE hInstance, HINSTANCE hPrevInstance, LPSTR lpCmdLine, int nShowCmd)
 {
-    // Trigger plugin onload event
-    s_framework->LoadPlugins();
+    s_framework->TriggerOnWinMain();
 
     return g_WinMain_hook.call<int>(hInstance, hPrevInstance, lpCmdLine, nShowCmd);
 }
 
+__declspec(noinline) void* hookedMhMainCtor(void* this_ptr)
+{
+    s_framework->TriggerOnMhMainCtor();
+
+    return g_MhMainCtor_hook.call<void*>(this_ptr);
```

Comment:
Maybe call the ctor before invoking the event. That way all singletons will have been instantiated already when the event is called.

Reply:
Good call - updated to do that.